node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1289,11 +1295,226 @@ async fn fetch_block_worker(
                 // want to retry.
                 error!(
                     ?err,
-                    ?block_hash,
+                    ?block_header_hash,
                     "failed to download block with infinite retries"
                 );
             }
         }
+
+        if let Some(finality_signatures) =
+            fetch_finality_signatures_by_block_header(block_header, ctx).await?
```

Comment:
In what case this method would return `None`? 

And also - it will return `Ok(Some(sigs))` for `FromStorage(_)` and in that case calling `put_signatures_to_storage` is unnecessary.

Reply:
> In what case this method would return None?

From what I understand if no signatures are found, `None` is returned. The signatures collector only initializes its inner set of signatures once at least one set of signatures is added.

Reply:
> In what case this method would return None?

In case there were no signatures fetched at all, i.e `BlockSignaturesCollector::add()` wasn't invoked.

> And also - it will return Ok(Some(sigs)) for FromStorage(_) and in that case calling put_signatures_to_storage is unnecessary.

I think the most clear way to address this is to change `fetch_finality_signatures_by_block_header` to `fetch_and_store_finality_signatures_by_block_header` and move the storing logic inside. Changes are here: c092138202fc956e9eba40b7b5d9cc9e68393876

Reply:
@fizyk20 @rafal-ch Looking at the `BlockSignaturesCollector`, it seems that it solely exists to make it possible to have a data structure to collect signatures with without specifying the block hash and era id --- otherwise I assume an empty `BlockSignatures` could suffice?

The only use of the collector seems to be in `fetch_finality_signatures_by_block_header`, but it has this information available. Is there merit to simply instantiation an empty set of signatures instead and using that?